extern/storage-sealing/commit_batch.go

```diff
@@ -114,6 +118,79 @@ func (b *CommitBatcher) run() {
 	}
 }
 
+func (b *CommitBatcher) batchWait(maxWait, slack time.Duration) time.Duration {
+	now := time.Now()
+
+	b.lk.Lock()
+	defer b.lk.Unlock()
+
+	var deadline time.Time
+	for sn := range b.todo {
+		sectorDeadline := b.getSectorDeadline(sn)
+		if deadline.IsZero() || (!sectorDeadline.IsZero() && sectorDeadline.Before(deadline)) {
+			deadline = sectorDeadline
+		}
+	}
+	for sn := range b.waiting {
+		sectorDeadline := b.getSectorDeadline(sn)
+		if deadline.IsZero() || (!sectorDeadline.IsZero() && sectorDeadline.Before(deadline)) {
+			deadline = sectorDeadline
+		}
+	}
+
+	if deadline.IsZero() {
+		return maxWait
+	}
+
+	deadline = deadline.Add(-slack)
+	if deadline.Before(now) {
+		return time.Nanosecond // can't return 0
+	}
+
+	wait := deadline.Sub(now)
+	if wait > maxWait {
+		wait = maxWait
+	}
+
+	return wait
+}
+
+func (b *CommitBatcher) getSectorDeadline(sn abi.SectorNumber) time.Time {
```

Comment:
I'd cache those as we can have 100s of sectors each with 100s of pieces, this will be doing a lot of RPC queries on each loop

Reply:
totally.